lib/ts/recipe/multifactorauth/recipe.ts

```diff
@@ -0,0 +1,508 @@
+    ): { status: "OK" } | { status: "VALIDATION_ERROR"; reason: string } => {
+        // this is called during sign in operations to ensure the secondary factor user is linked to the session user
+        // we disallow factor setup by sign in, and is allowed only via sign up
+        if (sessionUser.id !== factorUser.id) {
+            return {
+                status: "VALIDATION_ERROR",
+                reason: "Cannot complete MFA because of security reasons. Please contact support. (ERR_CODE_009)",
+            };
+        }
+
+        return {
+            status: "OK",
+        };
+    };
+
+    checkAllowedToSetupFactorElseThrowInvalidClaimError = async (
+        tenantId: string,
+        session: SessionContainerInterface,
+        sessionUser: User,
+        factorId: string,
+        userContext: UserContext
+    ): Promise<void> => {
+        // this is a utility function that populates all the necessary info for the recipe function
+        const tenantInfo = await Multitenancy.getTenant(tenantId, userContext);
+        if (tenantInfo === undefined) {
+            throw new SessionError({
+                type: SessionError.UNAUTHORISED,
+                message: "Tenant not found",
+            });
+        }
+
+        const requiredSecondaryFactorsForUser = await this.recipeInterfaceImpl.getRequiredSecondaryFactorsForUser({
+            userId: sessionUser.id,
+            userContext: userContext,
+        });
+        const factorsSetUpForUser = await this.recipeInterfaceImpl.getFactorsSetupForUser({
+            user: sessionUser,
+            userContext: userContext,
+        });
+        const completedFactorsClaimValue = await session.getClaimValue(MultiFactorAuthClaim, userContext);
+        const mfaRequirementsForAuth = await this.recipeInterfaceImpl.getMFARequirementsForAuth({
+            user: sessionUser,
+            accessTokenPayload: session.getAccessTokenPayload(userContext),
+            tenantId: tenantId,
+            factorsSetUpForUser,
+            requiredSecondaryFactorsForTenant: tenantInfo.requiredSecondaryFactors ?? [],
+            requiredSecondaryFactorsForUser,
+            completedFactors: completedFactorsClaimValue?.c ?? {},
+            userContext: userContext,
+        });
+
+        await this.recipeInterfaceImpl.checkAllowedToSetupFactorElseThrowInvalidClaimError({
+            session: session,
+            factorId: factorId,
+            completedFactors: completedFactorsClaimValue?.c ?? {},
+            requiredSecondaryFactorsForTenant: tenantInfo.requiredSecondaryFactors ?? [],
+            requiredSecondaryFactorsForUser,
+            factorsSetUpForUser,
+            mfaRequirementsForAuth,
+            userContext: userContext,
+        });
+    };
+
+    checkFactorUserAccountInfoForVerification = (
+        sessionUser: User,
+        accountInfo: { email?: string; phoneNumber?: string }
+    ): { status: "OK" } | { status: "VALIDATION_ERROR"; reason: string } => {
```

Comment:
why not just return a boolean from this function? and then have the API throw the right error.. we should not assume the context of this function call!

Reply:
Done